File: server/spec/factories/transactions.rb
# == Schema Information
#
# Table name: transactions
#
#  id                   :bigint           not null, primary key
#  amount               :decimal(10, 2)   not null
#  description          :text
#  executed_at          :datetime         not null
#  frequency            :string           default("one_time"), not null
#  transaction_type     :string           default("expense"), not null
#  created_at           :datetime         not null
#  updated_at           :datetime         not null
#  account_id           :bigint           not null
#  category_id          :bigint           default(0), not null
#  recipient_account_id :bigint
#  user_id              :bigint           not null
#
# Indexes
#
#  index_transactions_on_account_id            (account_id)
#  index_transactions_on_category_id           (category_id)
#  index_transactions_on_frequency             (frequency)
#  index_transactions_on_recipient_account_id  (recipient_account_id)
#  index_transactions_on_transaction_type      (transaction_type)
#  index_transactions_on_user_id               (user_id)
#
# Foreign Keys
#
#  fk_rails_...  (account_id => accounts.id)
#  fk_rails_...  (category_id => categories.id)
#  fk_rails_...  (recipient_account_id => accounts.id)
#  fk_rails_...  (user_id => users.id)
#
FactoryBot.define do
  factory :transaction do
    association :user
    association :account
    amount { Faker::Number.decimal(l_digits: 3, r_digits: 2) }
    transaction_type { Transaction.transaction_types[:expense] }
    description { Faker::Lorem.sentence }
    executed_at { DateTime.current }

    trait :transfer do
      transaction_type { Transaction.transaction_types[:transfer] }
      association :recipient_account, factory: :account
    end

    trait :income do
      transaction_type { Transaction.transaction_types[:income] }
    end

    trait :expense do
      transaction_type { Transaction.transaction_types[:expense] }
    end

    trait :executed_in_current_month do
      executed_at { 1.day.ago }
    end

    trait :executed_in_past_month do
      executed_at { 40.days.ago }
    end
  end
end

File: client/src/lib/types/transactions.ts
import { Account } from "./accounts";
import { TransactionCategory } from "./transaction_categories";
export interface Transaction {
  id: string;
  account_id: string;
  recipient_account_id?: string;
  amount: number;
  transaction_type: string;
  description: string;
  executed_at: string;
  frequency: string;
  account?: Account;
  created_at: string;
  updated_at: string;
  recipient_account_name?: string;
  recipient_account?: Account;
  category?: TransactionCategory;
}

export type TransactionType = string;

export interface TransactionsFilterParams {
  account_id?: string;
  transaction_type?: string;
  start_date?: string;
  end_date?: string;
  search?: string;
  frequency?: string;
  category?: string;
  page?: number;
  per_page?: number;
}

export interface CreateTransactionData {
  account_id: string;
  recipient_account_id?: string;
  amount: number;
  transaction_type: string;
  description?: string;
  executed_at?: string;
  frequency?: string;
  category_id?: string;
}